### ISSUE TITLE
Disable long outdated Nextcloud instances

### ISSUE THREAD

**Issue by MorrisJobke** [nextcloud/server#14230]
**Is your feature request related to a problem? Please describe.**

There are at least 8.000 instances out there that contact our updater server that have a version up to 13.0.0, 12.0.5, 11.0.7 or 10.0.6. All those versions have in common, that they were released more than a year ago. That means (except for 10.0.6, which was already end of life by then), that there was a minor release afterwards, that fixed something on that version - be it a normal misbehavior or a security bug.

As the main goal of Nextcloud is to be able to  host your data on your on servers and that comes with a big plus in privacy there is the downside of the need to manage updates on your own. There is no privacy without security.

**Describe the solution you'd like**

The idea is to have a staged escalation of behavior of the Nextcloud instance which ends with the complete isolation - feature wise - after a given date. This might sound hard, but let me explain it in more detail.

With staged I mean that there are different steps to make the admins and users aware of the potential issue. The whole idea bases on the release date of the given version that currently runs.

Let me put here an example of how this could work:

* 2018-03-15: release date of 13.0.1 that currently runs on this example instance
* 2018-09-15: all users in the group `admin` (and the group specified in the update notification app) are notified (via push notifications and email) that it is recommended to update - otherwise all users are notified about this 1 month later
* 2019-01-15: all users of the instance are notified that an update is recommended
* 2019-02-15: all users are notified, that the instance will go into safety mode if no update happens within the next month
* 2019-03-15: instances goes into safety mode

This safety mode means that the instance will not serve anything beside the login prompt that the admin can use to login (and only admins) to use the web updater to upgrade the instance. Maybe we just completely turn off the web UI with a nice warning and an update needs to be run manually.

Obviously the times are now just some example values, but 9 months for the admins, 10 months for all users, 11 months for all users with the warning of the safety mode and 12 months to trigger the safety mode seems to be a sane amount of time.

**Describe alternatives you've considered**

Currently there are already notifications about updates and updates are installed quite often, still there is a big amount of instances out there that runs really old software. Also as this is an open source product you could obviously patch out the check and run the software in an air gapped env for example for eternity.

Beside this it's just one part of many to increase the security of our product. There is also some vague plans to provide a reliable way to install security updates automatically, but this is something for a separate ticket.

@karlitschek @jospoortvliet @nickvergessen @blizzz @ChristophWurst @juliushaertl @skjnldsv @jancborchardt @rullzer Opinions? Is this a good idea or something totally horrible and total no-go?

**Comment by karlitschek** [comment]
I agree that this is a real problem. Your proposal makes sense. But maybe we start with more aggressive notifications first. Big red waring on the login page and other things before we remotely kill an instance. :-)

**Comment by MorrisJobke** [comment]
> I agree that this is a real problem. Your proposal makes sense. But maybe we start with more aggressive notifications first. Big red waring on the login page and other things before we remotely kill an instance. :-)

Yep - makes sense as well 👍 Also it's good to start with those easier ones first 👍

**Comment by rullzer** [comment]
We can indeed do it step by step.
I also like the added notifications. I'd say, Every 3 months (so roughly 2 patch releases) a reminder to update to the whole admin team is also fair. We could make the wording more urgent with each step.

**Comment by arnowelzel** [comment]
I would agree that having outdated instances can be a problem for security. But completely disabling access may lead to the situation that user will not trust Nextcloud any longer since they always fear to get locked out if they don't keep their instances updated. Even worse: people might consider keeping the latest Nextcloud version without this feature running just to avoid that lockout.

So I also would vote for a reminder to the admin team and some visual indication as well in the UI which can't be removed - maybe a warning with white text in front of a red background above the toolbar etc..

**Comment by nickvergessen** [comment]
The main issue I see, are people that do not use the Web interface at all. I guess this is also the majority of affected instances. In those cases, I think the sync client should come with similar prompts at some point.

**Comment by rullzer** [comment]
The sync client show notifications ;-)

**Comment by j-ed** [comment]
I think if you ask people why they are not updating the software you will hear the same answers as for other software packages, like e.g.

* I need to stick to PHP 5.6 because there is a legacy app installed on that server which is still required.
* I've installed a packaged Nextcloud on my NAS and the package is not maintained anymore.
* The server has been installed by a colleague who has left the company and there is no-one who knows how an upgrade is done.
* Why should I upgrade it is still working.
* I don't have time to upgrade the software.

I think you will never manage to get all installations updated, except you're going to disable support of older versions in the available sync clients, because people will usually update their mobile sync clients more often than the server software.

**Comment by bpcurse** [comment]
Some general issues in the update process need to be cleared prior to this:
- Not all apps are compatible with new versions and it sometimes takes very long until this is fixed. For 15 it is atm e.g. limit login by ip or the passman issue with the failing integrity check. With a growing app store this probably won't get better and many different apps might be mission critical in the diverse nc use cases.
- Updates in shared hosting environments with nextcloud one-click install that put the temp directory inside nextcloud root (e.g. all-inkl) are impossible for users with low technical skills as the updater won't work due to the additional directory.
Either there should be an override option during the gui update process or better the updater should accept a temp directory inside nc root that is defined in config.php.

**Comment by nickvergessen** [comment]
@bpcurse while I understand your reasons, I think you can also play it the other way around.
An old/not-updated app should not be a reason blocking people from upgrading.
If developers have problems to update their app for the next release, that is another thing to look into.

**Note** we are not talking about an aggressive strategy here, so let me calculate an example for one of the apps you mentioned. So you are currently bound to Nextcloud 14 due to "limit login by ip":

* Nextcloud 14.0.0 was released on 2018-09-10, so the suggested notifications for 14.0.0 would start in ~4 months only 2019-06-10.
* Also you can update to 14.0.7 which was released 11 days ago on 2019-02-07, so there would be notifications in 2019-11-07 only.
* End of life for Nextcloud 14 is planned for September this year 2019-09-XY, so if we do a release there and you upgrade to it, you will not get a notification before 2020-06-XY, although you run an out dated Nextcloud by then which is not receiving any security updates anymore, so I would really recommend looking an alternative for the app or to drop the app until an update is available.

So only after 21 months of the 14 release, the missing "limit login by ip" is a problem which will make you run into these notifications.

Just some more background:
Nextcloud 14.0.0 was released with support for PHP 7.0-7.2 
By 2020-06-XY:
* PHP7.0 is not supported by the PHP Group at all since 1.5 years (it ran End-Of-Life in January 2019)
* PHP7.1 is not supported by the PHP Group at all since 5-6 months (it is currently in security fixes only and runs End-Of-Life in January 2020)
* PHP7.2 is not actively supported by the PHP Group anymore since 5-6 months (it runs out of support in January 2020 and is security only until January 2021)

http://php.net/supported-versions.php

As sad as it sounds, but you still run a publicly accessible piece of software on the internet. **Not installing any updates for more than 12 months, must not be considered okay**, as it puts all your data on risk.

**Comment by feltel** [comment]
I´m a long time ownCloud and now Nextcloud user and admin. I´d like to comment on the issue from a user or admin perspective. I understand the issue and support the idea of having the most recent code running whereever possible. But there could be valid reasons in not running the most recent version or patch level. It should be up to the admin to decide which version he wants to run. Also locking out users from accessing their data when running an outdated version is a no go in my eyes. I don´t like the idea of having someone to control if my NC instance is worth running or not. Its like a remote kill switch. 

There are millions of users out there which are not technically skilled and denying access to their files could lead to fear and doubt regarding reliability and trust into Nextcloud.

**Comment by MorrisJobke** [comment]
> Also locking out users from accessing their data when running an outdated version is a no go in my eyes. I don´t like the idea of having someone to control if my NC instance is worth running or not. Its like a remote kill switch.
> 
> There are millions of users out there which are not technically skilled and denying access to their files could lead to fear and doubt regarding reliability and trust into Nextcloud.

Just to highlight what I had in mind:

* adding the build day to the version.php
* then one year after that disallow any request (also potential attackers - because only very little code is executed before this check it should be a good prevention)
* if you really really want to run it longer, then you can update the version.php and add "9999-01-01" as build date and then you can run your Nextcloud for the next ~8000 years without the kill switch to take account

This should be a safety mechanism against forgotten servers out there and not to cut any freedoms. And as always: Nextcloud is open source and one could easily follow how this mechanism works (the relevant code is then linked in this ticket here and can be manipulated to your own as you like).

Does that make sense to you, @feltel?

**Comment by rullzer** [comment]
> But there could be valid reasons in not running the most recent version or patch level.

@feltel yes and no. Latest version sure there can be reasons you can't upgrade your environment right away. Latest patch level. Not really. Sure you might run x.y.z-1 for a few weeks before you hit a maintenance window. But running x.y.0 when x.y.8 is out. There is no real excuse for that.

Same as if you are still running a NC 10 server. You are putting things at risk. Same way I would argue that somebody still running a debian 5 Server is not being a responsible admin.

I think we will first go with the aggressive notifications anyway to see if that helps.

**Comment by feltel** [comment]
@MorrisJobke okay, after your explanation I think I´m fine with this. As long as the behaviour is admin-controllable (maybe a setting in config.php which defaults to true) it should be not that much of a problem. But reading your initial proposal the first time one must think there sould be some kind of central instance where the "allowed" versions are noted and a Nextcloud installation checks regulary against this if it is allowed to be accessible. 

@rullzer But there could be appliances out there (different NAS-boxes come to mind; around 2016 there was a "Nextcloud Box" with NC11 iirc)  with pre-installed Nextcloud where there is no active development anymore or any time in the future. Why should it not allowed to run such boxes which are often not accessible through the internet?

Just to make it clear: I support the "running the latest version whenever possible"-statement but forcing admins to take action by locking out its users should not happen under any circumstands.

**Comment by Spartachetto** [comment]
I do understand the security concerns that guide this proposal. I also share the doubts of @feltel .

I will tell you how I'd think the process to be, maybe it'll give some ideas:
- when the Nextcloud instance is not updated from _x_ months (or: its security rating drops to _y_) start to notify the administrator of the security risk.
- after _y_ more months increase the notifications to the administrator and notify also the users (that could be unaware of the risk)
- you could consider a third step: **_blocking the access to federated clouds_ ** . 

The third step has, according to me, some advantages:
- the users of the instance keep their right of using whatever software they want to use
- the block has a good justification in securing users of other instances.

A softer approach would be to block the federation with **up to date instances** maintaining the possibility to federated with other insecure instances

**Comment by alexanderdd** [comment]
I propose to implement the first part of this idea, i.e. "Warning: your version will not be supported after Feb 2019" https://github.com/nextcloud/server/issues/15174

**Comment by alexanderdd** [comment]
I propose to implement the first part of this idea which seems to have wide support from people here, i.e. "Warning: your version will not be supported after Feb 2019" https://github.com/nextcloud/server/issues/15174

The second, more intrusive part seems to need some more discussion.

**Comment by J0WI** [comment]
+1 for all notification enhancements for admins.

> 2019-01-15: all users of the instance are notified that an update is recommended
> 2019-02-15: all users are notified, that the instance will go into safety mode if no update happens within the next month

This sounds horrible to me. A private instance could be used by a huge amount of users that have no clue about NextCloud. Such a warning would not be useful for them and may generate a significant load on mail servers/inboxes.

> 2019-03-15: instances goes into safety mode

This kind of a remote "kill switch" is really bad. They may be various reasons for not updating, admins may apply their own patches or the instance runs in a private network.
I would follow the principle that admins are responsible for their own services.

I would rather prefer automated updates (opt-out), just like WordPress. If the server is not maintained and the update fails, it will stuck in maintenance mode which should also be "safe".

**Comment by skjnldsv** [comment]
Closing?

**Comment by rullzer** [comment]
@skjnldsv nah we should still do the agressive notifications. Let me see if I can schedule some time for that.

**Comment by ChristophWurst** [comment]
> They may be various reasons for not updating, admins may apply their own patches or the instance runs in a private network.
> I would follow the principle that admins are responsible for their own services.

To be fair if someone is capable of patching their instance for security bugs, they'll also be capable of patching out the EOL mechanism.

**Comment by skjnldsv** [comment]
From #15174 

Or maybe we can check when the first beta is released for the upcoming major and start displaying a warning?

Like when 22 beta 1 is released, all 19 show a warning stating

> The versions is about to reach end of life, please upgrade to a more recent version!

**Comment by alexanderdd** [comment]
**Notifications** 
"..is about to reach end of life" is better than at the moment, but I opened https://github.com/nextcloud/server/issues/15174 because I think there should be more exact information on **when** a release will reach end of life. I don't think calculations like "after a year" are good because release schedule might change, also "when beta comes out" does not sound reliable enough for timing end of life.

IMO:
1. minimal implementation: in settings > version info: link to this page https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule
2. better: next to version info, display when this version will reach end of life

**Auto-update:**
I also like this:
> I would rather prefer automated updates (opt-out), just like WordPress. If the server is not maintained and the update fails, it will stuck in maintenance mode which should also be "safe".

**Comment by tgoeg** [comment]
I understand the issue and mostly think it's a good thing to do as described.
Concerning notifications to "all users", try to think the other way round:
The goal here is to protect data (at least I guess so).
If 8000 servers times I don't know how many (normal, non-admin) users get a notification about the insecurity of the used instance, there might be a few potential attackers in this list of maybe a million users that are going to enjoy this information very much.
Don't expect every user to be a friendly user.
It might end up putting the data more at risk than to protect it.

**Comment by nickvergessen** [comment]
> there might be a few potential attackers in this list 

They can check on several list for outdated software and notice whether or not an instance is outdated in many ways.

**Comment by jospoortvliet** [comment]
Summarizing/creating a proposal:
* warn admins 2 weeks after a minor patch update is out and they haven't updated.
* warn admins again after the next patch update is out.
  * Add that users will see this on the login page soon
  * And that users will get notified
  * I would do this over doing it time-based, as we can change timing of releases. This is more connected with disclosure and risk, I think. For example if we did a quick security fix in between the usual cycle, the admins are nudged faster.
  * and it would push for updates it faster - what Morris proposed was a REALLY long time.
* put warning on login page after the next patch update is out and a week later no update is done (~2 months).
  * It is super easy to check (server)/version.php so we're not leaking anything new.
  * End of life servers don't get minor updates so there we can only use a time based approach - 2 months or so?
* to discuss: after a few more months we could put server in "safe mode" (limit login to admin)
  * this should then also be show on that login page warning...

**Comment by tgoeg** [comment]
* make it configurable in config.php (as outlined by @feltel and me for various reasons)

Don't get me wrong - I very much consider making regular updates is absolutely crucial when hosting services.
Just don't force people into situations they cannot get out of easily or - worse - actively spill information per mail that won't help (end users) in all situations.

**Comment by feltel** [comment]
I´m fine notifying/nagging the admin of an NC instance about missed patch updates. I don´t see what informing the users of an instance would help in such a situation. They don´t have the capacity/power/knowledge to solve the problem. Also putting pressure to the admin though its users is at least questionable.

I would like to raise one more point. There could be situations where upgrading NC to the next major/still supported version is not possible due to the lack of compatible app versions. We faced this situation several time in recent years. We are using user_sql to make use of an already existing user base/data. user_sql is not available trough the regular app update mechanics since a long time. I know and use the fork done by @rotdrop . I´m grateful for his work. So I´m able to use NC in an still supported version. What I want to say is that there may be valid use cases for not having the most recent NC version running and even situations where updating to such a version is impossible, when a core functionality (i.e. authenticating users) is no longer possible.

**Comment by nickvergessen** [comment]
> There could be situations where upgrading NC to the next major/still supported version is not possible due to the lack of compatible app versions.

This is not the major subject here. We are talking about "long outdated", not things that are still upcoming and actively maintained.

- 🔴 We are talking about things like Nextcloud 16.0.0 being run publicly accessible. There is no reason to not update to 16.0.11 which was released **4th June 2020**.
- 🔴 From my POV it's also not okay to run 16.0.11 publicly accessible, which is EOL since 5+ years by now.
- 🟢 We are not talking about "I'm still running 31.0.x and can't update to 32.0.x yet as app Y is missing". 31 is still actively maintained, so this is no problem at all.
- 🟡 When you would still be running 30.0.x which is EOL since end of last year, you should question this decision, as there are no security updates anymore and security fixes and advisories we publish for newer versions could make bad people curious and try to apply the attack vectors to older versions not explicitly listed. 

> user_sql is not available trough the regular app update mechanics since a long time. I know and use the fork done by @rotdrop . I´m grateful for his work.

If this is the case, @rotdrop could ask to take over the app as it's not maintained anymore. We can revoke the old certificate and grant a new one. We did this from time to time already when maintainers simply disappeared and never came back. In other cases the maintainer reacted and agreed on handing the maintainership over.
